lib/basedriver/image-helpers.js

```diff
@@ -0,0 +1,110 @@
+ * @param {Region} rect The selected region of image
+ * @return {string} base64 encoded string of cropped image
+ */
+async function cropBase64Image (base64Image, rect) {
+  const image = await base64ToImage(base64Image);
+  const croppedImage = cropImage(image, rect);
+  return await imageToBase64(croppedImage);
+}
+
+/**
+ * Create a pngjs image from given base64 image
+ *
+ * @param {string} base64Image The string with base64 encoded image
+ * @return {PNG} The image object
+ */
+async function base64ToImage (base64Image) {
+  const imageBuffer = new Buffer(base64Image, 'base64');
+  return new B((resolve, reject) => {
+    const image = new PNG({filterType: 4});
+    image.parse(imageBuffer, (err, image) => { // eslint-disable-line promise/prefer-await-to-callbacks
+      if (err) { return reject(err); }
+      resolve(image);
+    });
+  });
+}
+
+/**
+ * Create a base64 string for given image object
+ *
+ * @param {PNG} image The image object
+ * @return {string} The string with base64 encoded image
+ */
+async function imageToBase64 (image) {
+  return new B((resolve, reject) => {
+    const chunks = [];
+    image.pack().on('data', (chunk) => chunks.push(chunk)).on('end', () => {
+      resolve(Buffer.concat(chunks).toString('base64'));
+    }).on('error', (err) => { // eslint-disable-line promise/prefer-await-to-callbacks
+      reject(err);
+    });
+  });
+}
+
+/**
+ * Crop the image by given rectangle
+ *
+ * @param {PNG} image The image to crop
+ * @param {Region} rect The selected region of image
+ * @return {PNG} Cropped image
+ */
+function cropImage (image, rect) {
+  const imageRect = {width: image.width, height: image.height};
+  const interRect = getRectIntersection(rect, imageRect);
+  if (isEmptyRect(interRect)) {
+    throw new Error(`Cannot crop ${JSON.stringify(rect)} from ${JSON.stringify(imageRect)} because the intersection between them is an empty rectangle`);
+  }
+
+  const firstVerticalPixel = interRect.top;
+  const lastVerticalPixel = interRect.top + interRect.height;
+
+  const firstHorizontalPixel = interRect.left;
+  const lastHorizontalPixel = interRect.left + interRect.width;
+
+  const croppedArray = [];
+  for (let y = firstVerticalPixel; y < lastVerticalPixel; y++) {
+    for (let x = firstHorizontalPixel; x < lastHorizontalPixel; x++) {
+      const firstByteIdxInPixelBlock = (imageRect.width * y + x) << 2;
+      for (let byteIdx = 0; byteIdx < BYTES_IN_PIXEL_BLOCK; byteIdx++) {
+        croppedArray.push(image.data[firstByteIdxInPixelBlock + byteIdx]);
+      }
+    }
+  }
+
+  image.data = new Buffer(croppedArray);
+  image.width = interRect.width;
+  image.height = interRect.height;
+  return image;
+}
+
+function getRectIntersection (rect, imageSize) {
+  const left = rect.left >= imageSize.width ? imageSize.width : rect.left;
+  const top = rect.top >= imageSize.height ? imageSize.height : rect.top;
+  const width = imageSize.width >= (left + rect.width) ? rect.width : 0;
```

Comment:
it looks like what this is doing is setting width to 0 even if there is a partial intersection. we want to disallow partial intersections? (rather than say resizing the crop rect to just go to the edge of the image?)

i don't care either way, just want to make sure we know what is going on. some comments might be useful here. or help me understand if i'm getting the logic wrong.

Reply:
Right now, the method accepts only regions that fully satisfied by given image. Otherwise, we set width (or height) to 0 and then throw an error (so, we use 0 as flag that something wrong with the region).

For the methods in which we use the method we don't need to support partial intersections, bacause we always know a size of image. If you think whould be better to support partial intersections, I can update the code.

Reply:
nope, no need! just making sure we were OK with this.